pkg/apis/serving/v1alpha1/build_compat.go

```diff
@@ -0,0 +1,101 @@
+/*
+Copyright 2018 The Knative Authors.
+
+Licensed under the Apache License, Version 2.0 (the "License");
+you may not use this file except in compliance with the License.
+You may obtain a copy of the License at
+
+    http://www.apache.org/licenses/LICENSE-2.0
+
+Unless required by applicable law or agreed to in writing, software
+distributed under the License is distributed on an "AS IS" BASIS,
+WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+See the License for the specific language governing permissions and
+limitations under the License.
+*/
+
+package v1alpha1
+
+import (
+	"bytes"
+	"encoding/json"
+	"errors"
+
+	"k8s.io/apimachinery/pkg/runtime"
+
+	buildv1alpha1 "github.com/knative/build/pkg/apis/build/v1alpha1"
+)
+
+// RawExtension is modeled after runtime.RawExtension, and should be
+// replaced with it (or an alias) once we can stop supporting embedded
+// BuildSpecs.
+type RawExtension struct {
+	// Field order is the precedence for JSON marshaling if multiple
+	// fields are set.
+	Raw       []byte
+	Object    runtime.Object
+	BuildSpec *buildv1alpha1.BuildSpec
+}
+
+var _ json.Unmarshaler = (*RawExtension)(nil)
+var _ json.Marshaler = (*RawExtension)(nil)
+
+func (re *RawExtension) UnmarshalJSON(in []byte) error {
+	if re == nil {
+		return errors.New("RawExtension: UnmarshalJSON on nil pointer")
+	}
+	if !bytes.Equal(in, []byte("null")) {
+		re.Raw = append(re.Raw[0:0], in...)
```

Comment:
Why not `re.Raw = in`?

Reply:
I copied this from `runtime.RawExtension`

Reply:
I think append is going to create a new slice, whereas assignment would create an alias.

Reply:
https://github.com/go101/go101/wiki/How-to-efficiently-clone-a-slice